File: src/kod/lexer/token.rs
use std::collections::HashMap;

#[derive(Debug, PartialEq, Clone)]
pub struct Token {
    #[allow(dead_code)]
    pub token_type: TokenType,
    #[allow(dead_code)]
    pub keyword_type: KeywordType,
    #[allow(dead_code)]
    pub value: String,
    #[allow(dead_code)]
    pub int_value: i64,
    #[allow(dead_code)]
    pub float_value: f64,
    #[allow(dead_code)]
    pub line: usize,
    #[allow(dead_code)]
    pub column: usize,
}

#[derive(Debug, PartialEq, Clone, Copy, Hash, Eq)]
pub enum TokenType {
    Unknown,
    EndOfFile,

    ADD,                //  +
    SUB,                //  -
    DIV,                //  /
    MUL,                //  *
    MOD,                //  %
    POW,                //  **

    AddEq,                //  +=
    SubEq,                //  -=
    DivEq,                //  /=
    MulEq,                //  *=
    ModEq,                //  %=

    AND,                //  &
    OR,                 //  |
    HAT,                //  ^
    SHL,                //  <<
    SHR,                //  >>
    NOT,                //  ~

    BoolNot,           //  !
    BoolEq,            //  ==
    BoolNe,            //  !=
    BoolLt,            //  <
    BoolGt,            //  >
    BoolLte,           //  <=
    BoolGte,           //  >=
    BoolAnd,           //  &&
    BoolOr,            //  ||
    ID,                 //  main x y foo
    KEYWORD,            //  NOT USED

    STRING,             //  "Hello world"
    INT,                //  5 6 456
    FLOAT,              //  6.9 7893.6   

    LPAREN,             //  (   
    RPAREN,             //  )              
    LBRACKET,           //  [              
    RBRACKET,           //  ]              
    LBRACE,             //  {          
    RBRACE,             //  }

    EQUALS,             //  =   
    COMMA,              //  ,  
    DOT,                //  .  
    COLON,              //  :  
    NAMESPACE,          //  ::  
    SEMI,               //  ;   
    QUESTION,           //  ?   
    AT,                 //  @
    HASH,               //  #
    LineComment,       // //
    MultilineCommentStart,     // /*
    MultilineCommentEnd,       // */
    POINTER,            //  ->
    ARROW,              //  =>
    BACKSLASH,          // 

    NewLine,           //  \n
}

#[derive(Debug, PartialEq, Clone, Copy)]
pub enum KeywordType {
    Unknown,
    Null,
    If,
    Else,
    While,
    For,
    Return,
    Import,
    As,
    From,
    Break,
    Continue,
}

pub fn get_symbols() -> HashMap<&'static str, TokenType> {
    HashMap::from([
        ("::", TokenType::NAMESPACE),
        ("&&", TokenType::BoolAnd),
        ("||", TokenType::BoolOr),
        ("**", TokenType::POW),
        ("//", TokenType::LineComment),
        ("/*", TokenType::MultilineCommentStart),
        ("*/", TokenType::MultilineCommentEnd),
        ("->", TokenType::POINTER),
        ("!=", TokenType::BoolNe),
        ("==", TokenType::BoolEq),
        ("=>", TokenType::ARROW),
        ("<<", TokenType::SHL),
        (">>", TokenType::SHR),
        ("<=", TokenType::BoolLte),
        (">=", TokenType::BoolGte),
        ("(", TokenType::LPAREN),
        (")", TokenType::RPAREN),
        ("[", TokenType::LBRACKET),
        ("]", TokenType::RBRACKET),
        ("{", TokenType::LBRACE),
        ("}", TokenType::RBRACE),
        ("=", TokenType::EQUALS),
        (",", TokenType::COMMA),
        (":", TokenType::COLON),
        (";", TokenType::SEMI),
        ("?", TokenType::QUESTION),
        ("%", TokenType::MOD),
        ("\\", TokenType::BACKSLASH),
        ("#", TokenType::HASH),
        ("@", TokenType::AT),
        ("+", TokenType::ADD),
        ("-", TokenType::SUB),
        ("/", TokenType::DIV),
        ("*", TokenType::MUL),
        ("&", TokenType::AND),
        ("|", TokenType::OR),
        ("^", TokenType::HAT),
        ("<", TokenType::BoolLt),
        (">", TokenType::BoolGt),
        ("~", TokenType::NOT),
        ("!", TokenType::BoolNot),
        (".", TokenType::DOT),
        ("+=", TokenType::AddEq),
        ("-=", TokenType::SubEq),
        ("/=", TokenType::DivEq),
        ("*=", TokenType::MulEq),
        ("%=", TokenType::ModEq),
    ])
}

impl Token {
    pub fn new(type_: TokenType, value: String, line: usize, column: usize) -> Self {
        Token {
            token_type: type_,
            keyword_type: KeywordType::Unknown,
            value,
            int_value: 0,
            float_value: 0.0,
            line,
            column,
        }
    }

    // pub fn to_string(&self) -> String {
    //     // Implement the to_string method logic
    //     // You can use the format! macro or other string formatting options in Rust
    //     format!("Token: {:?}, Value: {:}, Line: {}, Column: {}", self.token_type, self.value, self.line, self.column)
    // }
}

pub fn eof() -> Token {
    Token {
        token_type: TokenType::EndOfFile,
        keyword_type: KeywordType::Unknown,
        value: String::new(),
        int_value: 0,
        float_value: 0.0,
        line: 0,
        column: 0,
    }
}

File: Dev/KodLib/Token.cpp
#include "Token.hpp"

namespace Kod
{
    Location::Location(const std::wstring& file_path, uint32_t line, uint32_t column) :
        m_file_path(file_path),
        m_line(line),
        m_column(column)
    {}

    std::wostream& operator<<(std::wostream& os, const Location& location)
    {
        os << location.m_file_path.c_str() << ":" << location.m_line << ":" << location.m_column;
        return os;
    }

    Token::Token(const Location location, const TokenType type, const std::wstring& value, const KeywordType keyword_type) :
        m_location(location),
        m_token_type(type),
        m_value(value),
        m_keyword_type(keyword_type),
        m_int_value(0),
        m_float_value(0.0f)
    {}

    TokenType Token::s_symbol_to_type(const std::wstring& symbol)
    {
        static const std::unordered_map<std::wstring, TokenType> symbol_to_type = {
            { L"::", TokenType::NAMESPACE },
            { L"&&", TokenType::BOOL_AND },
            { L"||", TokenType::BOOL_OR },
            { L"**", TokenType::POW },
            { L"//", TokenType::LINE_COMMENT },
            { L"/*", TokenType::MULTILINE_COMMENT_START },
            { L"*/", TokenType::MULTILINE_COMMENT_END },
            { L"->", TokenType::POINTER },
            { L"!=", TokenType::BOOL_NE },
            { L"==", TokenType::BOOL_EQ },
            { L"=>", TokenType::ARROW },
            { L"<<", TokenType::SHL },
            { L">>", TokenType::SHR },
            { L"<=", TokenType::BOOL_LTE },
            { L">=", TokenType::BOOL_GTE },
            { L"(", TokenType::LPAREN },
            { L")", TokenType::RPAREN },
            { L"[", TokenType::LBRACKET },
            { L"]", TokenType::RBRACKET },
            { L"{", TokenType::LBRACE },
            { L"}", TokenType::RBRACE },
            { L"=", TokenType::EQUALS },
            { L",", TokenType::COMMA },
            { L":", TokenType::COLON },
            { L";", TokenType::SEMI },
            { L"?", TokenType::QUESTION },
            { L"%", TokenType::MOD },
            { L"\\", TokenType::BACKSLASH },
            { L"#", TokenType::HASH },
            { L"@", TokenType::AT },
            { L"+", TokenType::PLUS },
            { L"-", TokenType::MINUS },
            { L"/", TokenType::DIV },
            { L"*", TokenType::MUL },
            { L"&", TokenType::AND },
            { L"|", TokenType::OR },
            { L"^", TokenType::HAT },
            { L"<", TokenType::BOOL_LT },
            { L">", TokenType::BOOL_GT },
            { L"~", TokenType::NOT },
            { L"!", TokenType::BOOL_NOT },
            { L".", TokenType::DOT },
            { L"+=", TokenType::PLUS_EQ },
            { L"-=", TokenType::MINUS_EQ },
            { L"/=", TokenType::DIV_EQ },
            { L"*=", TokenType::MUL_EQ },
            { L"%=", TokenType::MOD_EQ },
        };
        const auto iter = symbol_to_type.find(symbol);
        return symbol_to_type.end() == iter ? TokenType::UNKNOWN : (*iter).second;
    }

    KeywordType Token::s_keyword_to_type(const std::wstring& keyword)
    {
        static const std::unordered_map<std::wstring, KeywordType> keyword_to_type = {
            { L"null", KeywordType::NULL_K },
            { L"fn", KeywordType::FN },
            { L"if", KeywordType::IF },
            { L"else", KeywordType::ELSE },
            { L"while", KeywordType::WHILE },
            { L"for", KeywordType::FOR },
            { L"return", KeywordType::RETURN },
            { L"import", KeywordType::IMPORT },
            { L"as", KeywordType::AS },
            { L"from", KeywordType::FROM },
            { L"break", KeywordType::BREAK },
            { L"continue", KeywordType::CONTINUE },
            { L"true", KeywordType::TRUE },
            { L"false", KeywordType::FALSE },
        };
        const auto iter = keyword_to_type.find(keyword);
        return keyword_to_type.end() == iter ? KeywordType::UNKNOWN : (*iter).second;
    }

    bool Token::_s_is_symbol(wchar_t character)
    {
        static const std::wstring symbols(L"()[]{}=@#,.:;?\\+-/*%&|^<>!~");
        return std::wstring::npos != symbols.find(character);
    }

    std::wostream& operator<<(std::wostream& os, const Token& token)
    {
        os << token.m_location << ": (" << static_cast<uint32_t>(token.m_token_type) << "): " << token.m_value;
        return os;
    }
}
